src/Microsoft.DotNet.Build.Tasks/FilterRuntimesFromSupports.cs

```diff
@@ -0,0 +1,294 @@
+
+            List<CompatibilityProfile> filteredSupports = new List<CompatibilityProfile>();
+            List<CompatibilityProfile> filteredRuntimes = new List<CompatibilityProfile>();
+
+            if (Frameworks!=null && Frameworks.Length > 0)
+            {
+                foreach (var key in supportsGraph.Supports.Keys)
+                {
+                    var compat = DefaultCompatibilityProvider.Instance;
+                    CompatibilityProfile compatibilityProfile = null;
+                    supportsGraph.Supports.TryGetValue(key, out compatibilityProfile);
+                    bool isFrameworkFiltered = false;
+                    foreach (var tfmRidPair in compatibilityProfile.RestoreContexts)
+                    {
+                        foreach (string tfm in Frameworks)
+                        {
+                            if (compat.IsCompatible(NuGetFramework.Parse(tfm), tfmRidPair.Framework))
+                            {
+                                isFrameworkFiltered = true;
+                                filteredSupports.Add(compatibilityProfile);
+                                break;
+                            }
+                        }
+                        if (isFrameworkFiltered)
+                        {
+                            break;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                foreach (var key in supportsGraph.Supports.Keys)
+                {
+                    CompatibilityProfile compatibilityProfile = null;
+                    supportsGraph.Supports.TryGetValue(key, out compatibilityProfile);
+                    filteredSupports.Add(compatibilityProfile);
+                }
+            }
+
+            if (RuntimeIds.Length > 0)
+            {
+                foreach (var compatibilityProfile in filteredSupports)
+                {
+                    List<bool> toRemove = new List<bool>();
+                    List<FrameworkRuntimePair> frameworkRuntimePairs = new List<FrameworkRuntimePair>();
+                    foreach (var tfmRidPair in compatibilityProfile.RestoreContexts)
+                    {
+                        bool isRuntimeFiltered = false;
+                        foreach (string runtimeId in RuntimeIds)
+                        {
+                            if (runtimeGraph.AreCompatible(runtimeId, tfmRidPair.RuntimeIdentifier))
+                            {
+                                isRuntimeFiltered = true;
+                                break;
+                            }
+                        }
+                        toRemove.Add(isRuntimeFiltered);
+                    }
+                    for (int i = 0, k = 0; i < toRemove.Count; i++)
+                    {
+                        if (!toRemove[i])
+                        {
+                            compatibilityProfile.RestoreContexts.RemoveAt(k);
```

Comment:
You can avoid this boolean array by using a for loop over compatibilityProfile.RestoreContexts in reverse order and calling RemoveAt(i).